modules/publication-service-oaipmh/src/main/java/org/opencastproject/publication/oaipmh/OaiPmhPublicationServiceImpl.java

```diff
@@ -79,453 +80,648 @@
+        publication = updateMetadata(job, mediaPackage, job.getArguments().get(1), flavors, tags, checkAvailability);
+        break;
+      default:
+        throw new IllegalArgumentException("Can not handle this type of operation: " + job.getOperation());
+    }
+    return publication != null ? MediaPackageElementParser.getAsXml(publication) : null;
   }
 
-  /**
-   * Callback for the OSGi declarative services configuration.
-   *
-   * @param distributionService
-   *          the distribution service
-   */
-  public void setDownloadDistributionService(DownloadDistributionService distributionService) {
-    this.downloadDistributionService = distributionService;
-  }
+  @Override
+  public Job publish(MediaPackage mediaPackage, String repository, Set<String> downloadElementIds,
+          Set<String> streamingElementIds, boolean checkAvailability)
+          throws PublicationException, MediaPackageException {
+    if (mediaPackage == null)
+      throw new IllegalArgumentException("Media package must be specified");
+    if (StringUtils.isEmpty(repository))
+      throw new IllegalArgumentException("Repository must be specified");
+    if (!oaiPmhServerInfo.hasRepo(repository))
+      throw new IllegalArgumentException("OAI-PMH repository '" + repository + "' does not exist");
 
-  /**
-   * Callback for the OSGi declarative services configuration.
-   *
-   * @param distributionService
-   *          the distribution service
-   */
-  public void setStreamingDistributionService(StreamingDistributionService distributionService) {
-    this.streamingDistributionService = distributionService;
+    try {
+      return serviceRegistry.createJob(JOB_TYPE, Operation.Publish.toString(),
+              Arrays.asList(MediaPackageParser.getAsXml(mediaPackage), // 0
+                      repository, // 1
+                      StringUtils.join(downloadElementIds, SEPARATOR), // 2
+                      StringUtils.join(streamingElementIds, SEPARATOR), // 3
+                      Boolean.toString(checkAvailability))); // 4
+    } catch (ServiceRegistryException e) {
+      throw new PublicationException("Unable to create job", e);
+    }
   }
 
-  /**
-   * Callback for the OSGi declarative services configuration.
-   *
-   * @param persistence
-   *          the OAI-PMH persistence
-   */
-  public void setPersistence(OaiPmhDatabase persistence) {
-    this.persistence = persistence;
-  }
+  @Override
+  public Job retract(MediaPackage mediaPackage, String repository) throws PublicationException, NotFoundException {
+    if (mediaPackage == null)
+      throw new IllegalArgumentException("Media package must be specified");
 
-  /** OSGi DI. */
-  public void setOaiPmhServerInfo(OaiPmhServerInfo oaiPmhServerInfo) {
-    this.oaiPmhServerInfo = oaiPmhServerInfo;
-  }
+    if (StringUtils.isNotBlank(repository) && !oaiPmhServerInfo.hasRepo(repository))
```

Comment:
Please try to be consistent: The checks in the method publish do the exactly same thing but do it differently. May factor the common code into a method to ensure consistency?

Reply:
I will put the common checks in a extra method. Thank you.